### PR TITLE
[DF] Compatibility in RDataSource without exceptions

### PR DESCRIPTION
**Pull request by hahnjo**
As ROOT cannot catch JITted exceptions on Apple Silicon (see the bug report https://github.com/root-project/root/issues/7541 for context), some tests are currently failing there due to the use of exceptions in the compatibility code for `RDataSource`. Implement this with a boolean flag and a pattern inspired by `errno`.

### CODE REVIEW

**Comment by hahnjo** [comment]
FWIW, tested with
```diff
diff --git a/tree/dataframe/inc/ROOT/RTrivialDS.hxx b/tree/dataframe/inc/ROOT/RTrivialDS.hxx
index b8e72d9f71..bd36d232e1 100644
--- a/tree/dataframe/inc/ROOT/RTrivialDS.hxx
+++ b/tree/dataframe/inc/ROOT/RTrivialDS.hxx
@@ -46,7 +46,7 @@ public:
    std::vector<std::pair<ULong64_t, ULong64_t>> GetEntryRanges();
    bool SetEntry(unsigned int slot, ULong64_t entry);
    void SetNSlots(unsigned int nSlots);
-   void Initialize();
+   void Initialise();
    std::string GetLabel();
 };

diff --git a/tree/dataframe/src/RTrivialDS.cxx b/tree/dataframe/src/RTrivialDS.cxx
index 93710c5181..31919f98df 100644
--- a/tree/dataframe/src/RTrivialDS.cxx
+++ b/tree/dataframe/src/RTrivialDS.cxx
@@ -95,8 +95,9 @@ void RTrivialDS::SetNSlots(unsigned int nSlots)
    fCounterAddr.resize(fNSlots);
 }

-void RTrivialDS::Initialize()
+void RTrivialDS::Initialise()
 {
+   std::cout << "RTrivialDS::Initialise()" << std::endl;
    if (fSize == std::numeric_limits<ULong64_t>::max()) {
       // infinite source, nothing to do here
       return;
```
and `./tree/dataframe/test/dataframe_interface --gtest_filter=RDFHelpers.CastToNode`

File: tree/dataframe/inc/ROOT/RDataSource.hxx [hunk 1]
```diff
@@ -197,6 +197,7 @@ public:
    // clang-format on
    virtual void Initialize() {}
 
+private:
```
Comment (eguiraud) [inline tree/dataframe/inc/ROOT/RDataSource.hxx:200]: these should be private, but ROOT coding conventions then demand that these entities be lifted up to the beginning of the class definition (before the first `public:`). The `/// \cond` and `///\ endcond` need to be refactored accordingly

**Review by eguiraud — APPROVED** [review]
LGTM, just a minor coding convention thing. Thank you very much Jonas!

File: tree/dataframe/inc/ROOT/RDataSource.hxx (hunk 1 above)
Comment (hahnjo) [reply tree/dataframe/inc/ROOT/RDataSource.hxx]: Like this? Just learned about the `\cond` / `\endcond` stuff...

**Review by eguiraud — APPROVED** [review]
LGTM, thank you very much!

**Comment by hahnjo** [comment]
The fast builds succeeded, and `mac11` as well as `windows10` passed for the previous push so I'm sparing Jenkins that.

I guess we also want this for v6.26?

**Comment by eguiraud** [comment]
I don't think the deprecation is present in v6.26

**Comment by hahnjo** [comment]
> I don't think the deprecation is present in v6.26

Yeah, should have checked :+1:

**Pull request merged**